gossip: rely on sorted SystemConfig in SystemConfigDeltaFilter

nvb opened the pull request:
See https://github.com/cockroachdb/cockroach/pull/21473#issuecomment-358265693.

SystemConfigs are always sorted by key, so instead of maintaining a map of keys
in SystemConfigDeltaFilter, we can just hold onto the previous SystemConfig and
scan over the new and old configs in order to find new keys and modified values.

```
name                       old time/op    new time/op    delta
SystemConfigDeltaFilter-4     106µs ± 2%      37µs ± 2%   -65.50%  (p=0.000 n=10+10)

name                       old alloc/op   new alloc/op   delta
SystemConfigDeltaFilter-4    1.10kB ± 0%    0.00kB       -100.00%  (p=0.000 n=10+10)

name                       old allocs/op  new allocs/op  delta
SystemConfigDeltaFilter-4      40.0 ± 0%       0.0       -100.00%  (p=0.000 n=10+10)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21610)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I'm surprised there was that much of a speedup. 

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/gossip/util.go, line 54 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21610#-L3LDydXvFaVy6XZFhNX:-L3LDydXvFaVy6XZFhNY:b6rx7xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d6d8c1693a54abe47dec927c7b3bdc13f02a9a/pkg/gossip/util.go#L54)):*
> ```Go
> 	for _, newKV := range newCfg.Values {
> 		// Skip unimportant kvs.
> 		if df.keyPrefix != nil && !bytes.HasPrefix(newKV.Key, df.keyPrefix) {
> ```

You could use `sort.Search` to find the start index in `newCfg.Values` and `oldCfg.Values` rather the iterating unnecessarily. Similarly, you can exit the iteration as soon as the prefix no longer matches.

---

*[pkg/gossip/util.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21610#-L3LEunD6hP8TJRcWNOi:-L3LEunD6hP8TJRcWNOj:b-nxjhpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d6d8c1693a54abe47dec927c7b3bdc13f02a9a/pkg/gossip/util.go#L59)):*
> ```Go
> 
> 		// Find matching key in lastCfg.
> 		for {
> ```

The nested for-loop is a bit odd. Normally looping over 2 arrays in parallel uses a single for-loop which either advances an index in one array or another or both. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21610#-:-L3LETwymEXrmO2VSM4h:b-38wxs2)*
<!-- Sent from Reviewable.io -->

nvb committed:
gossip: rely on sorted SystemConfig in SystemConfigDeltaFilter

See https://github.com/cockroachdb/cockroach/pull/21473#issuecomment-358265693.

SystemConfigs are always sorted by key, so instead of maintaining a map of keys
in SystemConfigDeltaFilter, we can just hold onto the previous SystemConfig and
scan over the new and old configs in order to find new keys and modified values.

```
name                       old time/op    new time/op    delta
SystemConfigDeltaFilter-4     105µs ± 2%      31µs ± 2%   -70.28%  (p=0.000 n=10+10)

name                       old alloc/op   new alloc/op   delta
SystemConfigDeltaFilter-4    1.10kB ± 0%    0.00kB       -100.00%  (p=0.000 n=10+10)

name                       old allocs/op  new allocs/op  delta
SystemConfigDeltaFilter-4      40.0 ± 0%       0.0       -100.00%  (p=0.000 n=10+10)
```

Release note: None

Release note: None

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/gossip/util.go, line 54 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21610#-L3LDydXvFaVy6XZFhNX:-L3UfC62AXXEFD5M5Isj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d6d8c1693a54abe47dec927c7b3bdc13f02a9a/pkg/gossip/util.go#L54)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You could use `sort.Search` to find the start index in `newCfg.Values` and `oldCfg.Values` rather the iterating unnecessarily. Similarly, you can exit the iteration as soon as the prefix no longer matches.
</blockquote></details>

Done.

---

*[pkg/gossip/util.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21610#-L3LEunD6hP8TJRcWNOi:-L3UfCX2_FgRkSOypUTU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5d6d8c1693a54abe47dec927c7b3bdc13f02a9a/pkg/gossip/util.go#L59)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The nested for-loop is a bit odd. Normally looping over 2 arrays in parallel uses a single for-loop which either advances an index in one array or another or both. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21610)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21610#-:-L3Uh0BnmJwIiNXNDryz:bnfp4nl)*
<!-- Sent from Reviewable.io -->
